src/config/include/config/Node.hpp

```diff
@@ -0,0 +1,664 @@
+              typename... Args,
+              typename = std::enable_if_t<isNodeConstructible<O, Args...>()>>
+    std::optional<O> _maybeImpl(Args&&... args) const {
+        // If you get compiler errors in _maybeImpl, see the note on the public maybe() function.
+        return std::make_optional<O>(*this, std::forward<Args>(args)...);
+    }
+
+    template <typename O,
+              typename... Args,
+              typename = std::enable_if_t<
+                  // don't have an `O(const Node&, Args...)` constructor
+                  !isNodeConstructible<O, Args...>() &&
+                  // rely on sfinae to determine if we have a `NodeConvert<O>` definition.
+                  std::is_same_v<O, typename NodeConvert<O>::type> &&
+                  // a bit pedantic but also require that we can call
+                  // `NodeConvert<O>::convert` and get back an O.
+                  std::is_same_v<O,
+                                 decltype(NodeConvert<O>::convert(std::declval<Node>(),
+                                                                  std::declval<Args>()...))>>,
+              typename = void>
+    std::optional<O> _maybeImpl(Args&&... args) const {
+        // If you get compiler errors in _maybeImpl, see the note on the public maybe() function.
+        return std::make_optional<O>(NodeConvert<O>::convert(*this, std::forward<Args>(args)...));
+    }
+
+    // This will get used if the ↑ fails to instantiate due to SFINAE.
+    template <
+        typename O,
+        typename... Args,
+        typename = std::enable_if_t<
+            // don't have an `O(const Node&, Args...)` constructor
+            !isNodeConstructible<O, Args...>() &&
+            // and the `YAML::convert<O>` struct has been defined.
+            std::is_same_v<decltype(YAML::convert<O>::encode(std::declval<O>())), YAML::Node>>,
+        typename = void,
+        typename = void>
+    std::optional<O> _maybeImpl(Args&&... args) const {
+        static_assert(sizeof...(args) == 0,
+                      "Cannot pass additional args when using built-in YAML conversion");
+        return std::make_optional<O>(yaml().as<O>());
+    }
+
+    const std::unique_ptr<const class NodeImpl> _impl;
+};
+
+/**
+ * The `[k,v]` used when doing `for(auto& [k,v] : node)`
+ */
+class NodeIteratorValue final : public std::pair<const v1::NodeKey&, const Node&> {
+public:
+    NodeIteratorValue(const v1::NodeKey& key, const Node& node);
+};
+
+/**
+ * Iterates over nodes. See `Node::begin()` and `Node::end()`.
+ */
+class NodeIterator final {
+public:
+    ~NodeIterator();
+
+    bool operator!=(const NodeIterator&) const;
+
+    bool operator==(const NodeIterator&) const;
+
+    void operator++();
+
+    const NodeIteratorValue operator*() const;
+
+private:
+    friend Node;
+
+    NodeIterator(const class NodeImpl*, bool end);
+
+    const std::unique_ptr<class IteratorImpl> _impl;
+};
+
+
+// A bulk of this could probably be moved to the PiMPL version
```

Comment:
This makes sense. Add a thing to the GI doc?

Reply:
👍 GI-067